examples/platforms/samr21/radio.c

```diff
@@ -0,0 +1,611 @@
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements the OpenThread platform abstraction for radio communication.
+ *
+ */
+
+#include "asf.h"
+
+#include <openthread/config.h>
+#include <openthread/platform/alarm-milli.h>
+#include <openthread/platform/radio.h>
+
+#include "platform-samr21.h"
+#include "common/logging.hpp"
+
+enum
+{
+    IEEE802154_ACK_LENGTH      = 5,
+    IEEE802154_FCS_SIZE        = 2
+};
+
+static otRadioFrame sTransmitFrame;
+static uint8_t      sTransmitPsdu[OT_RADIO_FRAME_MAX_SIZE];
+
+static otRadioFrame sReceiveFrame;
+
+static bool         sSleep          = false;
+static bool         sRxEnable       = false;
+static bool         sTxDone         = false;
+static bool         sRxDone         = false;
+static otError      sTxStatus       = OT_ERROR_NONE;
+static int8_t       sPower          = OPENTHREAD_CONFIG_DEFAULT_MAX_TRANSMIT_POWER;
+static otRadioState sState          = OT_RADIO_STATE_DISABLED;
+static bool         sPromiscuous    = false;
+static uint8_t      sChannel        = 0xFF;
+
+static int8_t       sMaxRssi;
+static uint32_t     sScanStartTime;
+static uint16_t     sScanDuration;
+static bool         sStartScan      = false;
+
+static int8_t         sTxPowerTable[] = { 4, 4, 3, 3, 3, 2, 1, 0,
+                                          -1, -2, -3, -4, -6, -8, -12, -17
+                                        };
+
+/*******************************************************************************
+ * Static
+ ******************************************************************************/
+
+static void radioSleep()
+{
+    if (!sSleep)
+    {
+        PHY_SetRxState(false);
+        PHY_Sleep();
+
+        sSleep = true;
+        sRxEnable = false;
+    }
+}
+
+static void radioWakeup()
+{
+    if (sSleep)
+    {
+        PHY_Wakeup();
+    }
+}
+
+static void radioRxEnable()
+{
+    if (sSleep)
+    {
+        PHY_Wakeup();
+
+        sSleep = false;
+    }
+
+    if (!sRxEnable)
+    {
+        PHY_SetRxState(true);
+
+        sRxEnable = true;
+    }
+}
+
+static void radioTrxOff()
+{
+    if (sSleep)
+    {
+        PHY_Wakeup();
```

Comment:
do we need to set `sSleep = false` as well? another place is line 131.

Reply:
Yes, could be optimized in this way.